components/cloud-foundry/src/view/util/deploy-location/deploy-location.directive.js

```diff
@@ -0,0 +1,153 @@
+(function () {
+  'use strict';
+
+  angular
+    .module('cloud-foundry.view.applications.application')
+    .directive('deployLocation', deployLocation);
+
+  /**
+   * @memberof cloud-foundry.view.applications.application
+   * @name deployLocation
+   * @description
+   * @returns {object} The space-picker directive definition object
+   */
+  function deployLocation() {
+    return {
+      bindToController: true,
+      controller: DeployLocationController,
+      controllerAs: 'deployLocation',
+      templateUrl: 'plugins/cloud-foundry/view/util/deploy-location/deploy-location.html',
+      scope: {
+        serviceInstances: '=',
+        serviceInstance: '=?',
+        organization: '=?',
+        space: '=?'
+      }
+    };
+  }
+
+  /**
+   * @memberof cloud-foundry.view.applications.application
+   * @name DeployLocationController
+   * @constructor
+   * @param {object} $scope - Angular $scope
+   * @param {app.model.modelManager} modelManager - the Model management service
+   * @param {object} cfOrganizationModel - the cfOrganizationModel service
+   * @param {object} cfUtilsService - Utilities for cloud foundries
+   */
+  function DeployLocationController($scope, modelManager, cfOrganizationModel, cfUtilsService) {
+
+    var vm = this;
+
+    var appModel = modelManager.retrieve('cloud-foundry.model.application');
+    var authModel = modelManager.retrieve('cloud-foundry.model.auth');
+
+    vm.getOrganizations = getOrganizations;
+    vm.getSpacesForOrganization = getSpacesForOrganization;
+
+    init();
+
+    var initialServiceInstance = _.get(vm.serviceInstance, 'metadata.guid') || appModel.filterParams.cnsiGuid;
+    var initialOrganization = _.get(vm.organization, 'metadata.guid') || appModel.filterParams.orgGuid;
+    var initialSpace = _.get(vm.space, 'metadata.guid') || appModel.filterParams.spaceGuid;
+
+    function init() {
+      vm.organizations = [];
+      vm.spaces = [];
+
+      if (initialServiceInstance && initialServiceInstance !== 'all') {
+        // Find the option to set. If the user has no permissions this may be null
+        var preSelectedService = _.find(vm.serviceInstances, {value: {guid: initialServiceInstance}}) || {};
+        vm.serviceInstance = preSelectedService.value;
+      }
+
+      vm.stopWatchServiceInstance = $scope.$watch(function () {
```

Comment:
I don't see where stopWatchServiceInstance  is used? In the app app code we cleaned these up in the $destroy, but since they are $scope watches, they should get cleanup up for us automatically.